oak-blob-cloud-azure/src/main/java/org/apache/jackrabbit/oak/blob/cloud/azure/blobstorage/AzureBlobStoreBackend.java

```diff
@@ -0,0 +1,784 @@
+                        this,
+                        connectionString,
+                        containerName,
+                        new DataIdentifier(stripMetaKeyPrefix(blob.getName())),
+                        blob.getProperties().getLastModified().getTime(),
+                        blob.getProperties().getLength(),
+                        true));
+                }
+            }
+            LOG.debug("Metadata records read. recordsRead={} metadataFolder={} duration={}", records.size(), prefix, (System.currentTimeMillis() - start));
+        }
+        catch (StorageException e) {
+            LOG.info("Error reading all metadata records. metadataFolder={}", prefix, e);
+        }
+        catch (DataStoreException | URISyntaxException e) {
+            LOG.debug("Error reading all metadata records. metadataFolder={}", prefix, e);
+        }
+        finally {
+            if (null != contextClassLoader) {
+                Thread.currentThread().setContextClassLoader(contextClassLoader);
+            }
+        }
+        return records;
+    }
+
+    @Override
+    public boolean deleteMetadataRecord(String name) {
+        long start = System.currentTimeMillis();
+        ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+        try {
+            Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+
+            CloudBlockBlob blob = getAzureContainer().getBlockBlobReference(addMetaKeyPrefix(name));
+            boolean result = blob.deleteIfExists();
+            LOG.debug("Metadata record deleted. existed={} metadataName={} duration={}", result, name, (System.currentTimeMillis() - start));
+            return result;
+
+        }
+        catch (StorageException e) {
+            LOG.info("Error deleting metadata record. metadataName={}", name, e);
+        }
+        catch (DataStoreException | URISyntaxException e) {
+            LOG.debug("Error deleting metadata record. metadataName={}", name, e);
+        }
+        finally {
+            if (contextClassLoader != null) {
+                Thread.currentThread().setContextClassLoader(contextClassLoader);
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public void deleteAllMetadataRecords(String prefix) {
+        if (null == prefix) {
+            throw new NullPointerException("prefix");
+        }
+        long start = System.currentTimeMillis();
+        ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+        try {
+            Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+
+            CloudBlobDirectory metaDir = getAzureContainer().getDirectoryReference(META_DIR_NAME);
+            int total = 0;
+            for (ListBlobItem item : metaDir.listBlobs(prefix)) {
+                if (item instanceof CloudBlob) {
+                    ((CloudBlob)item).deleteIfExists();
+                    total ++;
```

Comment:
The variable "total" here is actually a count of all CloudBlob objects enumerated, and not a count of records deleted.  This is true because the deleteIfExists() call is used.  So if the current item does not exist, nothing will be deleted on that iteration through the loop, but the value of "total" will be incremented anyway.

Once this loop completes, a log message will be displayed indicating the number of records deleted, and the value of "total" is used to display that number.  But it is possible that the number may indicate more records were actually deleted if one or more blob items in the directory were deleted before but a reference still existed.

I don't know if that can actually happen or not; it depends on whether listBlobs() will return items that are actually deleted.  This was discussed in an earlier comment.

If it is possible that deleteIfExists() could return false, we should test the result of deleteIfExists() and only increment total if the result is true.  If this is not a possible case, a comment explaining this would help.